Define speculative HTML parsing

zcorpan opened the pull request:
Fixes #5624.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chromium (@chrishtr, @mfreed7)
   * Gecko (@hsivonen has [fixed a bug already](https://github.com/whatwg/html/issues/5624#issuecomment-689553341))
   * WebKit (@othermaciej's comments: https://github.com/whatwg/html/issues/5624#issuecomment-646966479 & https://github.com/whatwg/html/issues/5624#issuecomment-656072493)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/24521
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1144176
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1674583
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=229709

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5959/index.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">/index.html</a>  ( <a href="https://whatpr.org/html/5959/cdf62c8...3535735/index.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">diff</a> )
<a href="https://whatpr.org/html/5959/parsing.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">/parsing.html</a>  ( <a href="https://whatpr.org/html/5959/cdf62c8...3535735/parsing.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">diff</a> )
<a href="https://whatpr.org/html/5959/references.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">/references.html</a>  ( <a href="https://whatpr.org/html/5959/cdf62c8...3535735/references.html" title="Last updated on Sep 12, 2021, 7:50 PM UTC (3535735)">diff</a> )

chrishtr commented:
@domenic could you review, and @mfreed7 also?

hsivonen commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+
+    <p class="note">It is possible that the same markup is seen multiple times from the
+    <span>speculative HTML parser</span> and then the normal HTML parser. It is expected that
+    duplicated fetches will be prevented by normal caching rules.</p>
```
For better or worse, Gecko at present tries a speculative fetch at most once for each unique URL seen during a page load regardless of "normal caching rules".

zcorpan replied:
@hsivonen do you think the spec should suggest that strategy for speculative fetches?

zcorpan commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+  <ul>
+   <li><p>Let the HTML parser's invocation of the <span data-x="concept-appcache-init">application
+   cache selection algorithm</span> complete before initiating any speculative fetches, if
+   applicable.</p></li>
```
I hope appcache removal is successful so this can be removed... Though, this doesn't invoke appcache here, it says that if it had been invoked, wait for it to complete.

zcorpan replied:
Oh hey, #6153 removed appcache from the spec. So I remove this part now.

zcorpan commented:
@domenic @yoavweiss I've tried to address the feedback by being a bit more specific and inventing a concept of "speculative mock elements", which can't cause things to happen.

domenic commented on source:
```diff
@@ -112924,6 +112937,32 @@ document.body.appendChild(text);
+  <var>tagName</var>, and <var>attributes</var>:</p>
+
+  <ol>
+   <li><p>Let <var>element</var> be a new <span data-x="ordered map">map</span>.</p></li>
```
I'd probably define a speculative mock element as a struct having `<dfn>`ed namespace, local name, attributes, and child nodes, instead of using an ordered map this way. Since the keys are known, I don't think an ordered map is very appropriate.

zcorpan replied:
Done

domenic reviewed (commented):
Wow, this is pretty inventive. After starting at it for a bit, I think it works, and I like it!!

Still some other minor code review comments to resolve, but I think this is a very clean way of semi-rigorously building up a tree structure, while making it clear none of the normal mechanisms happen.

zcorpan commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+    fetches</span>. Which kinds of resources to speculatively fetch is
+    <span>implementation-defined</span>, but user agents must not speculatively fetch resources that
+    would not be fetched with the normal HTML parser, under the assumption that the script that is
+    blocking the HTML parser does nothing.</p>
```
Although I agree with avoiding double negatives generally, see 4c655f2bb71d99e2632ada3641813e12e8e8490a

zcorpan commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+   <var>speculativeParser</var>.</p></li>
+
+   <li><p><span>In parallel</span>, run <var>speculativeParser</var> until it is stopped or until it
+   reaches the end of its <span>input stream</span>.</p></li>
```
I think when the normal parser starts parsing again, the speculative parser is stopped. At least I think this is true for Gecko, less sure about Chromium and WebKit. cc @hsivonen @mfreed7 

In Gecko, I think a speculative parser can have a speculative parser, but I didn't do that in the spec to make the model a bit simpler. If a parser-blocking script `document.write`s another parser-blocking script, the spec starts over speculative parsing altogether. The main "win" for the spec is that it doesn't need to check whether the `document.write` did something that invalidates existing speculations. This can still be added in in the future if implementers would like to have that specified.

zcorpan commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+   </li>
+  </ul>
+
+  <p>The <dfn export>speculative fetches</dfn> must follow these rules:</p>
```
Removed `export`

zcorpan commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+  <ul>
+   <li><p>Scripts must not be executed.</p></li>
+
+   <li><p>Character encoding declarations must be ignored.</p></li>
```
Clarified in b71ba9bbdd933ac917859b0eb00f9df52cb57187 -- IIRC, this was based on feedback from @hsivonen during or around TPAC:
https://github.com/whatwg/html/issues/5624#issuecomment-718882519

Although "change the encoding" can in some cases cause a navigation (which would be bad to do speculatively), I think it would be fine to change the encoding for the speculative parser (without navigating, and without changing the encoding of the document "for real"). That is, I don't see anything in particular going wrong by doing so. However, I suspect that the positive impact of doing so is very close to zero, while the added implementation complexity is not.

zcorpan commented:
@domenic thanks!

Another reason for this approach is that e.g. the [Adoption Agency Algorithm](https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm) can mutate parts of the DOM tree before the parser-blocking script, which would be bad to do speculatively on the real DOM tree.

domenic commented on source:
```diff
@@ -112798,6 +112822,176 @@ document.body.appendChild(text);
+  <ul>
+   <li><p>Let the HTML parser's invocation of the <span data-x="concept-appcache-init">application
+   cache selection algorithm</span> complete before initiating any speculative fetches, if
+   applicable.</p></li>
```
Some appcache leftovers

domenic commented on source:
```diff
@@ -112798,6 +112822,176 @@ document.body.appendChild(text);
+    <p>If <var>parser</var> already has an <span>active speculative HTML parser</span>, then
+    <span>stop the speculative HTML parser</span> for <var>parser</var>.</p>
+
+    <p class="note">This can happen when <code data-x="">document.write()</code> writes another
```
Cross link document.write, i.e. fill in this data-x?

domenic commented on source:
```diff
@@ -112798,6 +112822,176 @@ document.body.appendChild(text);
+   to an element's <span data-x="concept-element-namespace">namespace</span>.</p></li>
+
+   <li><p>A <span>string</span> <dfn data-x="concept-mock-localName">localName</dfn>, corresponding
+   to an element's <span data-x="concept-element-local-name">local name</span>.</p></li>
```
I'd kind of prefer to just match the casing and names of element, i.e. "local name" and "children"?

domenic reviewed (commented):
Only a couple of minor things!

zcorpan commented:
@domenic thanks, fixed!

domenic commented on source:
```diff
@@ -112798,6 +112806,127 @@ document.body.appendChild(text);
+
+   <li>
+    <p>If the <span>speculative HTML parser</span> encounters one of the following elements, then
+    act as if that element is processed for the purpose of its effect of speculative fetches for
```
Tokens or tags seems slightly more correct from the spec perspective too; WDYT @zcorpan?

domenic reviewed (approved):
LGTM! It seems you've got a couple open threads left with @mfreed7 and @hsivonen; I'll let you decide whether you want to wait on resolving those or merge and potentially revisit later if they come back with feedback.

zcorpan commented:
I'd ideally like a LGTM from at least two implementers here.

andreubotella commented on source:
```diff
@@ -111441,6 +111465,173 @@ document.body.appendChild(text);
+
+  <ul>
+   <li>
+    <p>The state of the normal HTML parser and the document itself must not be affected.</p>
```
The behavior of the input stream and the input byte stream probably needs to be specified in more depth – since tokens [pushed into the input (byte) stream](https://html.spec.whatwg.org/multipage/browsing-the-web.html#read-html) must also be pushed into the speculative parser's stream, but tokens read from the streams should be independent.

zcorpan replied:
Thanks! https://github.com/whatwg/html/pull/5959/commits/6a20e1916ede7b84b368d11eb3eabff1344dae74

This list is hand-wavy, certainly, but the feedback from implementers is to not specify the speculative parser in precise detail as the approaches used in different engines are different and currently we're mostly concerned about observable differences and that this optimization is specified at all.

mfreed7 commented:
So overall I like this spec change. In essence, the preload scanner shouldn't load something that the "real" parser wouldn't eventually load also, assuming no `document.write()`'s change things in the meantime. As a spec, I think that makes sense, and due to Gecko's use of the actual parser / tree builder, most of the tests pass there. On the Chromium side, the implementation is quite separate, with the preload scanner being a very thin [approximation](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/html/parser/html_preload_scanner.cc;l=1014;drc=9a1bececa2156050162532e38fc15a183d036a12;bpv=1;bpt=1) of the real tree builder. As such, more of the tests fail, and some might be tough to fix. One alternative is to re-architect the preload scanner to use the real tree builder, but I don't think that's worth the effort. Which means the "fix" for Chromium will amount to trying to patch the holes in the approximations being made in the existing preload scanner. I'm ok with that. There's already [a nice bug](https://crbug.com/1144176) summarizing those errors for us.

TL;DR, I'm supportive of landing this PR along with the [test PR](https://wpt.fyi/results/html/syntax/speculative-parsing?label=pr_head&max-count=1&pr=24521).

annevk commented on source:
```diff
@@ -111441,6 +111465,173 @@ document.body.appendChild(text);
+   parsing.)</p></li>
+
+   <li>
+    <p>If <var>parser</var> already has an <span>active speculative HTML parser</span>, then
```
This should be a non-null check, right?

zcorpan replied:
https://github.com/whatwg/html/pull/5959/commits/f9119accb3b8865fb132194b201192e6a845d838

hsivonen commented:
Thanks for working on this.

The spec seems to allow speculative fetches only when a script is blocking the parser. In Firefox, speculative fetches can happen even when the parser is not blocked by script but DOM building actions are being accumulated into a larger batch of work and speculative fetches are started during that accumulation.

I think this is relevant to whether Firefox complies to the spec text in the as-if sense in case DOM changes come from other sources than parser-blocking scripts (e.g. async scripts or timeouts set by a same-origin parent). Specifically, if something else disconnects a node such that the parser keeps inserting nodes that are not in the document and this causes some non-speculative fetches not to occur, Firefox still performs speculative fetches. (Images are fetched even if disconnected, but I _think_ there are other fetch types where the non-speculative case requires the node to be in the document.)

I think it should be conforming to perform speculative fetches at any time on the assumption that no scripted action, regardless whether from a parser-blocking script or not, does anything.

In Firefox, duplicate request avoidance doesn't go all the way to the cache but the speculative load machinery refuses to speculatively fetch the same URL twice.

I don't see spec text connecting the creation of speculative mock elements to speculative fetches.

zcorpan commented:
@hsivonen 

> In Firefox, duplicate request avoidance doesn't go all the way to the cache but the speculative load machinery refuses to speculatively fetch the same URL twice.

Is it keyed off of only the URL, or a tuple of URL, `crossorigin` attribute's state, `referrerpolicy` attribute's state, or some such?

zcorpan commented:
> The spec seems to allow speculative fetches only when a script is blocking the parser. In Firefox, speculative fetches can happen even when the parser is not blocked by script but DOM building actions are being accumulated into a larger batch of work and speculative fetches are started during that accumulation.
> 
> I think this is relevant to whether Firefox complies to the spec text in the as-if sense in case DOM changes come from other sources than parser-blocking scripts (e.g. async scripts or timeouts set by a same-origin parent). Specifically, if something else disconnects a node such that the parser keeps inserting nodes that are not in the document and this causes some non-speculative fetches not to occur, Firefox still performs speculative fetches. [...]
> 
> I think it should be conforming to perform speculative fetches at any time on the assumption that no scripted action, regardless whether from a parser-blocking script or not, does anything.

I'm happy to allow it. I think we need to separate speculative parsing and speculative fetching a bit more in the spec, so normal parsing can also cause speculative fetches...

> (Images are fetched even if disconnected, but I _think_ there are other fetch types where the non-speculative case requires the node to be in the document.)

Yes, for example `<link rel=stylesheet>`, `<link rel=modulepreload>`.

`<video poster>` seems to not require the element to be [browsing-context connected](https://html.spec.whatwg.org/multipage/infrastructure.html#browsing-context-connected) per spec: https://html.spec.whatwg.org/multipage/media.html#poster-frame

The spec allows fetching `<script src>` when the `src` attribute is set, before the element is connected:

"For performance reasons, user agents may start fetching the classic script or module graph (as defined above) as soon as the src attribute is set, instead, in the hope that the element will be inserted into the document (and that the crossorigin attribute won't change value in the meantime)."
https://html.spec.whatwg.org/multipage/scripting.html#prepare-a-script

I can't tell for `<link rel=preload>`, SVG `<image>`.

> I don't see spec text connecting the creation of speculative mock elements to speculative fetches.

Yeah, this is maybe too handwavily implied currently. I'll try to address it along with allowing regular-parsing speculative fetches.

hsivonen commented:
>> In Firefox, duplicate request avoidance doesn't go all the way to the cache but the speculative load machinery refuses to speculatively fetch the same URL twice.
>
> Is it keyed off of only the URL, or a tuple of URL, crossorigin attribute's state, referrerpolicy attribute's state, or some such?

[URL only](https://searchfox.org/mozilla-central/rev/d3683dbb252506400c71256ef3994cdbdfb71ada/parser/html/nsHtml5TreeOpExecutor.cpp#1146) with the twist that if there is a media query that doesn't apply, the URL is ignored instead of listed as already loaded.

This might not be a good idea. This code predates the introduction of the `crossorigin` attribute and the `referrerpolicy` attribute.

zcorpan commented:
@hsivonen I've addressed your comments, except I haven't specified the media query twist. I specified the duplicate fetch prevention to be URL only for now.

zcorpan commented:
Arguably, "Let  url  be the  URL  that  element  would fetch if it was processed normally" covers the media query aspect (as well as `type` attribute, etc), though it's not very explicit.

hsivonen commented on source:
```diff
@@ -114123,6 +114157,191 @@ document.body.appendChild(text);
+
+  <p>User agents may implement an optimization, as described in this section, to speculatively fetch
+  resources that are declared in the HTML markup while the HTML parser is waiting for a
+  <span>pending parsing-blocking script</span> to be fetched and executed. While this optimization
```
I think limiting speculations to times where there is a pending parsing-blocking script is too limiting. Right now, I'm implementing speculation (assuming the inheritable encoding of same-origin parent or UTF-8) before `meta charset` has been seen. Moreover, in Firefox, even when the parsing doesn't diverge due to speculation, the speculative fetches can occur at points where there is no parsing-blocking script from the main-thread perspective.

zcorpan replied:
This paragraph needs updating to be accurate, since the spec now allows speculative fetches from normal parsing. Fixed in https://github.com/whatwg/html/pull/5959/commits/1be217dc0ab426e553a0976b402823da772e7106

I'm hesitant to specify speculative parsing during `meta charset` prescan before addressing https://github.com/whatwg/html/issues/6962 and https://github.com/whatwg/html/issues/5913 . Can we treat that as a separate issue, so that it doesn't block this PR?

hsivonen replied:
Let's treat it as a separate issue.

zcorpan replied:
https://github.com/whatwg/html/issues/7001

hsivonen commented on source:
```diff
@@ -114123,6 +114157,191 @@ document.body.appendChild(text);
+    fetch">speculative fetches</span>. Which kinds of resources to speculatively fetch is
+    <span>implementation-defined</span>, but user agents must not speculatively fetch resources that
+    would not be fetched with the normal HTML parser, under the assumption that the script that is
+    blocking the HTML parser does nothing.</p>
```
When in comes to speculating before `meta charset`, I'm in the process of extending this to under the assumption that the eventual encoding of the document will be the inherited one, if there is an inheritable encoding, or, if there is no inheritable encoding, UTF-8.

hsivonen reviewed (changes requested):
Thanks. I think the spec should allow speculation while looking for `meta charset` and the spec should allow fetches that the normal parser will see in the future to start speculatively. This could be explained by starting a parallel speculative parser at the start of the document with the HTTP-layer encoding, if there is one, or, otherwise, the inherited encoding, if there is one, or, otherwise, UTF-8.

mfreed7 commented on source:
```diff
@@ -114123,6 +114157,193 @@ document.body.appendChild(text);
+
+    <p class="note">It is possible that the same markup is seen multiple times from the
+    <span>speculative HTML parser</span> and then the normal HTML parser. It is expected that
+    duplicated fetches will be prevented by caching rules, which are not yet fully specified.</p>
```
Even if the caching rules aren't specified, do you think it would be possible to specify this normatively here? I.e. enforce that speculative fetches will **never** be followed by normal HTML parser fetches? If you generally agree, I'm ok if this is tackled as a follow-on PR.

zcorpan replied:
Yes, it would be good to require normatively, but might require some plumbing. Can you raise an issue?

mfreed7 replied:
https://github.com/whatwg/html/issues/7065

mfreed7 commented on source:
```diff
@@ -114123,6 +114157,193 @@ document.body.appendChild(text);
+
+  <ol>
+   <li><p>Optionally, return. (This allows user agents to opt out of speculative HTML
+   parsing.)</p></li>
```
nit: should the parenthetical here be a Note, instead of inline with this step?

zcorpan replied:
Fixed in https://github.com/whatwg/html/pull/5959/commits/3535735891b51346a43780dea64d1fa83e6aefe2

mfreed7 commented on source:
```diff
@@ -114123,6 +114157,193 @@ document.body.appendChild(text);
+  <p>When the tree builder says to insert an element into a <code>template</code> element's
+  <span>template contents</span>, if that is a <span>speculative mock element</span>, instead do
+  nothing. URLs found speculatively inside <code>template</code> elements might themselves be
+  templates, and must not be speculatively fetched.</p>
```
While the [declarative Shadow DOM spec](https://github.com/whatwg/dom/pull/892) isn't landed, I'm concerned that this will be incompatible. `<template>` elements that are declarative shadow roots should *likely* get speculatively parsed, since they will be added to the live DOM tree upon the closing `</template>` tag.

Mainly this is a note to myself to come back here to update the DSD PR, but let me know what you think.

zcorpan replied:
That seems fine. There's an attribute to denote declarative shadow DOM (`shadowroot`), correct? Would it be OK to speculatively fetch in those cases, but not for regular `template` elements?

mfreed7 replied:
Right, the attribute is `shadowroot`, and the legal values are `<template shadowroot=open>` or `<template shadowroot=closed>`. Anything other than those two values should look like a regular `<template>` and speculative parsers should skip.

zcorpan replied:
Filed #7069

mfreed7 reviewed (commented):
This looks pretty good to me. I like the "speculative mock element" concept - makes the spec pretty straightforward here. I added a few small comments, but overall LGTM.

zcorpan commented on source:
```diff
@@ -114123,6 +114157,193 @@ document.body.appendChild(text);
+  <span>pending parsing-blocking script</span> to be fetched and executed, or during normal parsing,
+  at the time <span data-x="create an element for the token">an element is created for a token</span>.
+  While this optimization is not defined in precise detail, there are some rules to consider for
+  nteroperability.</p>
```
Thanks, fixed in https://github.com/whatwg/html/pull/5959/commits/3535735891b51346a43780dea64d1fa83e6aefe2
